### ISSUE TITLE
How to fix this incompatible Flow types, without creating different methods

### ISSUE THREAD

**Issue by MichaelDeBoey** [facebook/flow#4978]
So I'm having the following situation (using `flow-bin: ^0.55.0` in my `package.json`):

I have an API config file on my server, which I want to `fetch`.
When it's a successful `fetch` (when it has an `OK` status), I return the response, else I throw the response.
If that's successful, I parse it as a JSON.
But Flow says the returned type is incompatible with the one I say it will return.
```js
Error: src/services/api/index.js:62
                v------------------------------------
 62:     return fetch(`${publicUrl}/config/api.json`)
 63:         .then(isSuccessful)
 64:         .then(asJson);
             ------------^ type application of class `Promise`. This type is incompatible with the expected return type of
 61: const getAPIConfig = (): Promise<APIConfig> => {
                              ^^^^^^^^^^^^^^^^^^ type application of identifier `Promise`
  Type argument `R` is incompatible:
     19: const asJson = (response: Response): JSON => {
                                              ^^^^ object type. This type is incompatible with
     61: const getAPIConfig = (): Promise<APIConfig> => {
                                          ^^^^^^^^^ object type
      Property `baseUrl` is incompatible:
          5:     [key: string]: string | JSON,
                                         ^^^^ object type. This type is incompatible with
         18:         [env: string]: string,
                                    ^^^^^^ string. See: src/flowtypes/index.js:18
```

Can you guys help me with an easy (and recommended) way to solve this one? (can also be seen [on flow.org/try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAJgUwMYwIYBOmYUArgHbYAuAlnOSZldgBZgAUAzgFxidUEa5AOYBKXgAUCcALY1OmADx5yATwB8Abgw58RMFVUAHYgCVMnI-QVgAvGADeqMGDgBrXgCM4cGJhUANM5gAFac9LzsorbqYCqqQQC+qCmGJmAAghIAkgDC9FA0wnaOwS6eeAoAqgQwvE4ujY0A2pjkAG68-IIiALpdAkLCAWDAwGD+rBMdNNLkMm1UYCyVcWDoNFBQmETkSxXVtWVgiUnaacQAUgDKAPIAciUNYM1umKoDPcL9fIMiYAAfMA3B5nFLYaxLYRMLJ5ApFEpRSTSOQKRSw-LkQrCWIxUpNIhUUgEBjbZgsdgAcmAEKxRWAeCMNAAdGF6JTRMcmsyqCw2ux5NdSNhsBZOGQYJymtzefzKpdwuRRNpEtpafwwILhaLOOLSDBEURLNZMLxzMbyApxGBzVZLcQ8c8XJsOEa7Qpme5RPjpS5CcSGG6TdomokJjAbE7pbzpEgg-aQ41kqrUOqlvLFYaLO7TTbsybrSDHo7jv6SWB4x62eQoirtEA))

```js
type APIConfig = {
  baseUrl: {
    [env: string]: string, // each environment has a different baseUrl
  },
};
type JSON = {
  [key: string]: string | JSON,
};


const getAPIConfig = (): Promise<APIConfig> => {
  return fetch('/config/api.json')
    .then(isSuccessful)
    .then(asJson);
};


const isSuccessful = (response: Response): Response => {
  if (response.ok) {
    return response;
  } else {
    throw response;
  }
};
const asJson = (response: Response): JSON => {
    return response.json();
};
```

Also adding a `asAPIConfig` method (see [flow.org/try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAJgUwMYwIYBOmYUArgHbYAuAlnOSZldgBZgAUAzgFxidUEa5AOYBKXgAUCcALY1OmADx5yATwB8Abgw58RMFVUAHYgCVMnI-QVgAvGADeqMGDgBrXgCM4cGJhUANM5gAFac9LzsorbqYCqqQQC+qCmGJmAAghIAkgDC9FA0wnaOwZ54CgCqBDC8Ti4uANqY5ABuvPyCIgC6HQJCwgFgwMBg-qxjbTTS5DItVGAsFXFg6DRQUJhE5AvlVTXBiUnaacQAUgDKAPIAciX1YI1umKp9XcK9fP0iYAA+YJdbscUthrAthEwsnkCkUSlFJNI5ApFFD8uRCsJYjFSi4iFRSAQGJtmCx2AByYCg9FFYB4Iw0AB0YXoZNEwRcDKoLBa7HkF1I2GwFk4ZBgbIaYE53PI7AqqJhYm0iW0VP4YD5AqFnBFpBgcKIlmsmF45kN5AU4jApqs5uI2Ie6w4BptCgZ7lEOIaeIJDGdRu0LkSYxgNgeLi50iQfttAbAyWVqFVCwqZ3CDHs7GjChNFhdmEtgLu9uC3sJYCzmCZaaiSpVYLinHl1OKGYrObNFt4TYxdliD1LDFlnFT9EzuaNlu7RVEtdQQA)) isn't working 😕
```js
const getAPIConfig = (): Promise<APIConfig> => {
  return fetch('/config/api.json')
    .then(isSuccessful)
    .then(asAPIConfig);
};

const asAPIConfig = (response: Response): APIConfig => {
  return (asJson(response): APIConfig);
};
```

**Comment by julienw** [comment]
Lazy but naughty one:
```javascript
const asAPIConfig = (response: Response): APIConfig => {
  return ((asJson(response): any): APIConfig);
};
```
But I think you should try to validate a little more your config. After all, the goal of flow is also to help you with this task.

On a side note, I noted on the try that
```javascript
type JSON = {
  [key: string]: string | JSON,
};
```
is really understood liek
```javascript
type JSON = {
  [key: string]: string | any,
};
```
Likely because of the infinite recursivity ?

**Comment by MichaelDeBoey** [comment]
@julienw The this is I **know** that the returned JSON from `'/config/api.json'` is in the `APIConfig` type, so I don't need to check that.
I just want to typecheck the rest of my code, so I want to say to `flow`: `"This _**is**_ something of 'APIConfig'"`

**Comment by jcready** [comment]
@julienw the `type-at-pos` reporting recursive types incorrectly using `any` is a known issue, see #4358. A bit unrelated, but a "proper" type for parsed JSON would be:

```js
type JSONValue = JSONObject | JSONArray | string | number | boolean | null
type JSONObject = { [key:string]: JSONValue }
type JSONArray = JSONValue[]
```

**Comment by jcready** [comment]
@MichaelDeBoey the type returned by `response.json()` is [already set to `Promise<any>`](https://github.com/facebook/flow/blob/fe5e0e3932b6743c8b62b40718d41a1cc1656a37/lib/bom.js#L963), so if you just remove your return annotation from `asJson` then you won't get any errors ([Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAJgUwMYwIYBOmYUArgHbYAuAlnOSZldgBZgAUAzgFxidUEa5AOYBKXgAUCcALY1OmADx5yATwB8Abgw58RMFVUAHYgCVMnI-QVgAvGADeqMGDgBrXgCM4cGJhUANM5gAFac9LzsorbqYCqqQQC+qCmGJmAAghIAkgDC9FA0wnaOwS6eeAoAqgQwvE4ujY0A2pjkAG68-IIiALpdAkLCAWDAwGD+rBMdNNLkMm1UYCyVcWDoNFBQmETkSxXVtWVgiUnaacQAUgDKAPIAciUNYM1umKoDPcL9fIMiYAAfMA3B5nFLYaxLYRMLJ5ApFEpRSTSOQKRSw-LkQrCWIxUpNIhUUgEBjbZgsdgAcmAEKxRWAeCMNAAdGF6JTRMcmsyqCw2ux5NdSNhsBZOGQYJymtzefzKpdwuRRNpEtpafwwILhaLOOLSDBEURLNZMLxzMbyApxGBzVZLcQ8c8XJsOEa7Qpme5RPjpS5CcSGG6TdomokJjAbE7pbzpEgg-aQ41kqrUOqlvLFYaLO7TTbsybvY7jv6SWB4x62eQoirtEA)):

```js
const asJson = (response: Response) => {
    return response.json();
};
```

**Comment by MichaelDeBoey** [comment]
@jcready I didn't know that was already in the standard `flow` defs. 😕
To make my code consistent, I'll provide the `Promise<any>` as the return type of `asJson` then. 🙂

**Comment by jcready** [comment]
@MichaelDeBoey I could caution you against annotating everything. One of the advantages of flow is it's ability to inter types for you, and some of the time your annotation might actually provide less specific types than flow would've inferred automatically for you.

**Comment by MichaelDeBoey** [comment]
@jcready Thanks for the warning, will take that in mind 🙂
